src/lib/EmissaryLib.sol

```diff
@@ -1,97 +1,210 @@
+     * when removing an emissary to keep storage clean and avoid stale data.
+     * @param sponsor The address of the sponsor
+     * @param allocator The address of the allocator
+     * @param newEmissary The address of the new emissary (use address(0) to remove)
+     * @param resetPeriod The reset period for assignment cooldown
+     * @param scope The scope of the assignment
+     */
+    function assignEmissary(address sponsor, address allocator, address newEmissary, ResetPeriod resetPeriod, Scope scope) internal {
+        bytes12 lockTag = allocator.toAllocatorIdIfRegistered().toLockTag(scope, resetPeriod);
+        EmissaryConfig storage config = _getEmissaryConfig(sponsor, lockTag);
+        uint256 _assignableAt = config.assignableAt;
+        address _emissary = config.emissary;
+
+        // Ensures that the assignment has been scheduled properly.
+        // Without this check, an emissary could be assigned without proper scheduling,
+        // leading to uncontrolled state transitions.
+        require(_assignableAt != NOT_SCHEDULED, EmissaryAssignmentUnavailable(NOT_SCHEDULED));
+
+        // The second check ensures that either there is no current emissary,
+        // or the reset period has elapsed before allowing a new assignment.
+        // This enforces the cooldown period between assignments to prevent abuse.
+        require(_emissary == address(0) || config.assignableAt <= block.timestamp, EmissaryAssignmentUnavailable(config.assignableAt));
+
+        // if new Emissary is address(0), that means that the sponsor wants to remove the emissary feature.
+        // we wipe all storage
+        if (newEmissary == address(0)) {
+            // If the new emissary is address(0), this means we are removing the emissary.
+            // We clear all related storage fields to maintain a clean state and avoid stale data.
+            delete config.emissary;
+            delete config.assignableAt;
+            delete config.resetPeriod;
         }
-
-        return _registerFor(sponsor, claimHashesAndTypehashes);
-    }
-
-    function assignEmissary(address sponsor, address emissary, uint256 nonce, uint256 expires, bool assigned, bytes calldata /*sponsorSignature*/ ) external returns (bool) {
-        expires.later();
-
-        bytes32 messageHash;
-        assembly ("memory-safe") {
-            let m := mload(0x40) // Grab the free memory pointer; memory will be left dirtied.
-            mstore(m, EMISSARY_ASSIGNMENT_TYPEHASH)
-            mstore(add(m, 0x20), emissary)
-            mstore(add(m, 0x40), nonce)
-            mstore(add(m, 0x60), expires)
-            mstore(add(m, 0x80), assigned)
-            messageHash := keccak256(m, 0xa0)
+        // otherwise we set the provided resetPeriod
+        else {
+            config.emissary = newEmissary;
+            config.assignableAt = NOT_SCHEDULED;
+            config.resetPeriod = resetPeriod;
```

Comment:
we don't need to store the reset period as we can extract it from the lock tag; we can also expand the assignableAt to a uint96 accordingly